src/DataStreams/PushingToViewsBlockOutputStream.cpp

```diff
@@ -169,39 +221,32 @@ void PushingToViewsBlockOutputStream::write(const Block & block)
             output->write(block);
     }
 
-    /// Don't process materialized views if this block is duplicate
-    if (!getContext()->getSettingsRef().deduplicate_blocks_in_dependent_materialized_views && replicated_output && replicated_output->lastBlockIsDuplicate())
+    if (views.empty())
         return;
 
-    // Insert data into materialized views only after successful insert into main table
+    /// Don't process materialized views if this block is duplicate
     const Settings & settings = getContext()->getSettingsRef();
-    if (settings.parallel_view_processing && views.size() > 1)
+    if (!settings.deduplicate_blocks_in_dependent_materialized_views && replicated_output && replicated_output->lastBlockIsDuplicate())
+        return;
+
+    const size_t max_threads = std::min(views.size(), (settings.parallel_view_processing ? static_cast<size_t>(settings.max_threads) : 1));
+    if (max_threads > 1)
     {
-        // Push to views concurrently if enabled and more than one view is attached
-        ThreadPool pool(std::min(size_t(settings.max_threads), views.size()));
+        ThreadPool pool(max_threads);
         for (auto & view : views)
         {
-            auto thread_group = CurrentThread::getGroup();
-            pool.scheduleOrThrowOnError([=, &view, this]
-            {
+            pool.scheduleOrThrowOnError([&] {
```

Comment:
I think that attaching a thread from a pool to a thread group is significant, because otherwise it will work without memory tracker.

Reply:
Or it already attached, because `ThreadStatus` object is stored inside a `view` object? This is kind of tricky.

Reply:
It will be done inside processXXX functions by switching to the ThreadStatus associated with each view (which have a memory profiler each).